connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSourceTask.java

```diff
@@ -322,22 +337,18 @@ private boolean sendRecords() {
                             @Override
                             public void onCompletion(RecordMetadata recordMetadata, Exception e) {
                                 if (e != null) {
-                                    // Given the default settings for zero data loss, this should basically never happen --
-                                    // between "infinite" retries, indefinite blocking on full buffers, and "infinite" request
-                                    // timeouts, callbacks with exceptions should never be invoked in practice. If the
-                                    // user overrode these settings, the best we can do is notify them of the failure via
-                                    // logging.
                                     log.error("{} failed to send record to {}: {}", WorkerSourceTask.this, topic, e);
                                     log.debug("{} Failed record: {}", WorkerSourceTask.this, preTransformRecord);
+                                    producerSendException.compareAndSet(null, e);
```

Comment:
@C0urante I am wondering if we should incorporate the `producer.send()` as a step in the source pipeline that must be handled by the `RetryAndToleranceOperator`. This can allow you to skip certain errors (for example, the RecordTooLargeException) which can be useful in certain use cases.

Reply:
That's a fascinating idea, @wicknicks. I like it and would be happy to take a stab at implementing it, but first I think we should clarify if that qualifies as a bug fix (which this PR is intended to be so that it can be backported to older branches) or a feature addition. I'm leaning slightly toward the latter but if others feel differently we can add it to this PR as well.

Reply:
+1 for a separate PR.